src/lib/utils/lenient-matching.ts

```diff
@@ -0,0 +1,35 @@
+import { type AnyZodObject, ZodObject, type ZodRawShape } from "zod";
+import { isMergeableObject } from '@/lib/utils/is-mergeable-object';
+
+/**
+ * Returns the data, with all the nested keys transformed to match the shape of the schema.
+ */
+export function applyLenientMatching(
+  data: Record<string, unknown>,
+  shape: ZodRawShape,
+): Record<string, unknown> {
+  return Object.fromEntries(
+    Object.entries(data).map(([key, value]) => {
+      const matchedKey =
+        Object.keys(shape).find(
+          (it) =>
+            it.replace(/[^a-z0-9]/gi, "").toLowerCase() ===
+            key.replace(/[^a-z0-9]/gi, "").toLowerCase(),
+        ) ?? key;
+
+        const zodType = shape[matchedKey];
+        if (isMergeableObject(value) && isZodObject(zodType)) {
+        return [
+          matchedKey,
+          applyLenientMatching(value, zodType.shape),
+        ];
+      } else {
+        return [matchedKey, value];
+      }
+    }),
+  );
+}
+
+function isZodObject(input: unknown): input is AnyZodObject {
+  return input instanceof ZodObject;
+}
```

Comment:
New utility, this basically deep-transforms the data to have the object keys match the corresponding shapes of the Zod schema.

By just having a lower-case alphanumerical comparison, it should match any naming convention difference and keeps this implementation detail away from the public API.

Reply:
If so, can we make the `applyLenientMatching` into something line `ApplyKeyMatching` where we can pass 1 or more matchers?

Reply:
Before proceeding, I'd like to point out that exposing the KeyMatcher function in the the public API might be unnecessary, and maybe expose to much implementation detail.

I know it was my own suggestion to begin with when I created the ticket, but I now believe that this lenient matching behavior in 99% of the cases is exactly what you want. 

What about only exposing it as an enum (union) option, so we allow ourselves to tweak this behavior later on without breaking the API?

e.g.
```ts
type Config = {
  // ...

  keyMatching: "lenient" | "strict"
}
```

This could then later be extended (with for example "case-insensitive") if there is a good use case where "lenient" is too lenient and strict too strict?

Reply:
I agree 👍  I think ` keyMatching: "lenient" | "strict"` is good. I prefer to keep it "strict" by default now and in the next major version let's switch it to "lenient" as you also mentioned.

Reply:
Awesome!! I will make the changes today or tomorrow 😁

Reply:
I pushed the changes which introduces an internal `KeyMatcher` system. Externally it remains as a simple enum (union).

Reply:
All good on this regards, thanks!